File: quick/golang/quickPessoa.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type Pessoa struct {
	codigo int
	name   string
}

func QuickSort(values []Pessoa, began, end int, comp, swap *uint64) []Pessoa {
	i := began
	j := end - 1
	pivo := values[(began+end)/2]

	for i <= j {
		for values[i].codigo < pivo.codigo || (values[i].codigo == pivo.codigo && values[i].name < pivo.name) {
			i++
			*comp += 3
		}
		for values[j].codigo > pivo.codigo || (values[j].codigo == pivo.codigo && values[j].name > pivo.name) {
			j--
			*comp += 3
		}
		if i <= j {
			values[i], values[j] = values[j], values[i]
			*swap++
			i++
			j--
		}
	}

	if j > began {
		QuickSort(values, began, j+1, comp, swap)
	}
	if i < end {
		QuickSort(values, i, end, comp, swap)
	}
	return values
}

func main() {
	input, err := os.Open("pessoa 1.txt")
	if err != nil {
		panic(err)
	}
	defer input.Close()
	output, err := os.OpenFile("./quickGoPessoa.txt", os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		panic(err)
	}
	defer input.Close()

	vetor := make([]Pessoa, 0)
	var comp, swap uint64

	num, err := strconv.Atoi(os.Args[1])

	scanner := bufio.NewScanner(input)
	for i := 0; i < num; i++ {

		scanner.Scan()
		linha := scanner.Text()
		valores := strings.Split(linha, "\t")
		id, _ := strconv.Atoi(valores[0])
		aux := Pessoa{id, valores[1]}
		vetor = append(vetor, aux)
	}


	inicio := time.Now()
	QuickSort(vetor, 0, num, &comp, &swap)
	fim := time.Now()


	fmt.Fprintf(output, "%s,%v,%d,%d\n", os.Args[1], float64(fim.Sub(inicio).Seconds()), comp, swap)
}

File: quick/cpp/quickPessoa.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <time.h>
#include <string>
#include <sstream>
using namespace std;

class Pessoa
{
public:
    int codigo;
    string name;
    Pessoa(){};
    Pessoa(int cod, string nome)
    {
        codigo = cod;
        name = nome;
    }
    void PrintPessoa()
    {
        cout << codigo << ' ' << name << endl;
    }
    bool operator<(Pessoa b)
    {
        return codigo < b.codigo;
    }
};

template <typename T>
void quicksort(vector<T> &values, int began, int end, unsigned long &comp, unsigned long &swap)
{
    int i, j;
    T pivo, aux;
    i = began;
    j = end - 1;
    pivo = values[(began + end) / 2];
    while (i <= j)
    {
        while (values[i].name < pivo.name && i < end)
        {
            i++;
        }
        while (values[j].name > pivo.name && j > began)
        {
            j--;
        }
        comp++;
        if (i <= j)
        {
            aux = values[i];
            values[i] = values[j];
            values[j] = aux;
            swap++;
            i++;
            j--;
        }
    }
    if (j > began)
        quicksort(values, began, j + 1, comp, swap);
    if (i < end)
        quicksort(values, i, end, comp, swap);
}

int main(int argc, char *argv[2])
{
    ifstream input("pessoa 1.txt");
    ofstream output("./quickCppPessoa.txt", ios_base::app);
    vector<Pessoa> vetor;
    unsigned long comp = 0, swap = 0;
    clock_t inicio, fim;

    for (int i = 0; i < atoi(argv[1]); i++)
    {
        string linha;

        getline(input, linha);
        stringstream ss(linha);
        string id_str, nome;

        getline(ss, id_str, '\t');
        getline(ss, nome, '\t');

        int id = stoi(id_str);
        vetor.push_back(Pessoa(id, nome));
    }

    inicio = clock();

    quicksort(vetor, 0, size(vetor), comp, swap);

    fim = clock();
    output << argv[1] << "," << (float)(((fim - inicio) + 0.0) / CLOCKS_PER_SEC) << "," << comp << "," << swap << "\n";

    return 0;
}